beacon-chain/state/state-native/setters_withdrawal.go

```diff
@@ -64,23 +67,74 @@ func (b *BeaconState) SetNextWithdrawalIndex(i uint64) error {
 		return errNotSupported("SetNextWithdrawalIndex", b.version)
 	}
 
-	b.lock.RLock()
-	defer b.lock.RUnlock()
+	b.lock.Lock()
+	defer b.lock.Unlock()
 
 	b.nextWithdrawalIndex = i
 	return nil
 }
 
+// IncreaseNextWithdrawalIndex increases the index that will be assigned to the next withdrawal.
+func (b *BeaconState) IncreaseNextWithdrawalIndex() error {
+	if b.version < version.Capella {
+		return errNotSupported("IncreaseNextWithdrawalIndex", b.version)
+	}
+
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.nextWithdrawalIndex += 1
+	return nil
+}
+
 // SetNextPartialWithdrawalValidatorIndex sets the index of the validator which is
 // next in line for a partial withdrawal.
 func (b *BeaconState) SetNextPartialWithdrawalValidatorIndex(i types.ValidatorIndex) error {
 	if b.version < version.Capella {
 		return errNotSupported("SetNextPartialWithdrawalValidatorIndex", b.version)
 	}
 
-	b.lock.RLock()
-	defer b.lock.RUnlock()
+	b.lock.Lock()
+	defer b.lock.Unlock()
 
 	b.nextPartialWithdrawalValidatorIndex = i
 	return nil
 }
+
+// WithdrawBalance withdraws the balance from the validator and creates and
+// withdrawal receipt for the EL to process
+func (b *BeaconState) WithdrawBalance(index types.ValidatorIndex, amount uint64) error {
+	if b.version < version.Capella {
+		return errNotSupported("WithdrawBalance", b.version)
+	}
+
+	if err := helpers.DecreaseBalance(b, index, amount); err != nil {
+		return errors.Wrap(err, "could not withdraw balance from validator")
```

Comment:
these error messages seems a bit generic, I'd recommend either return the error as it is or make it more descriptive like
 - could not decrease balance for index X
 - could not get the next withdrawal index
 - could not get validator at index X

Reply:
I don't like that pattern since I'm wrapping already an error that includes these

Reply:
I think the level is wrong here, `could not withdraw balance from validator` sounds more like it's coming from the caller of `WithdrawBalance`